File: ejemplos/nerve/09-probador-de-carga/ejemplo.py
"""
Ejemplo: Concurrencia con asyncio en Python

Qué enseña este ejemplo:
  - Usar `asyncio.gather` para lanzar múltiples tareas al mismo tiempo.
  - Medir cuánto tarda un bloque de código en ejecutarse usando `time.time()`.

Para tu reto (Probador de Carga Nerve):
  Reemplazarás `simular_peticion()` con `cliente.publicar(...)` o
  una llamada remota de Nerve para bombardear la red y medir el
  rendimiento total (mensajes por segundo).
"""

import asyncio
import time


# 1. Definir una tarea asíncrona que simula trabajo
async def simular_peticion(id_peticion: int):
    # Simulamos que enviar un mensaje tarda 100 milisegundos (0.1s)
    # Importante usar asyncio.sleep para no bloquear el hilo principal
    await asyncio.sleep(0.1)
    # Podrías retornar el resultado o simplemente confirmar
    return True


async def main():
    total_mensajes = 1000
    print(f"🚀 Iniciando prueba de carga: {total_mensajes} peticiones...")

    # Guardar el tiempo de inicio
    inicio = time.time()

    # 2. Crear una lista con todas las "promesas" o "tareas" a ejecutar
    tareas = []
    for i in range(total_mensajes):
        tareas.append(simular_peticion(i))

    # 3. asyncio.gather ejecuta todas las tareas en paralelo (concurrentemente)
    # Espera hasta que todas terminen y devuelve una lista con sus resultados
    resultados = await asyncio.gather(*tareas)

    # Calcular el tiempo total transcurrido
    fin = time.time()
    tiempo_total = fin - inicio

    exitosos = sum(1 for r in resultados if r is True)

    # 4. Mostrar estadísticas
    print(f"✅ Prueba finalizada en {tiempo_total:.2f} segundos.")
    print(f"📊 Mensajes exitosos: {exitosos}/{total_mensajes}")
    print(f"⚡ Rendimiento: {(total_mensajes / tiempo_total):.2f} peticiones/segundo")


if __name__ == "__main__":
    # Iniciar el bucle de eventos asíncrono
    asyncio.run(main())

File: ejemplos/nerve/09-probador-de-carga/ejemplo.go
/**
 * Ejemplo: Concurrencia con Goroutines y WaitGroups en Go
 * 
 * Qué enseña este ejemplo:
 *   - Lanzar "goroutines" (hilos ligeros) usando la palabra clave `go`.
 *   - Sincronizar las goroutines usando un `sync.WaitGroup` para esperar a que todas terminen.
 *   - Medir el tiempo transcurrido usando `time.Since`.
 * 
 * Para tu reto (Probador de Carga Nerve):
 *   Reemplazarás `simularPeticion()` con un envío al servidor usando 
 *   tu cliente Nerve para ver cuánto tráfico puede aguantar tu red.
 */

package main

import (
	"fmt"
	"sync"
	"time"
)

// 1. Tarea que se ejecutará de forma concurrente
// Recibe un puntero al WaitGroup para avisar cuando termine
func simularPeticion(idPeticion int, wg *sync.WaitGroup) {
	// Importante: al terminar la función, avisa al WaitGroup
	defer wg.Done()

	// Simulamos que enviar un mensaje tarda 100 milisegundos (0.1s)
	time.Sleep(100 * time.Millisecond)

	// Aquí iría tu cliente Nerve: clienteNerve.Publicar(...)
}

func main() {
	totalMensajes := 1000
	fmt.Printf("🚀 Iniciando prueba de carga: %d peticiones...\n", totalMensajes)

	// Guardar el tiempo de inicio
	inicio := time.Now()

	// 2. Crear un WaitGroup para esperar a que todas las goroutines terminen
	var wg sync.WaitGroup

	// Decirle al WaitGroup cuántas tareas vamos a lanzar
	wg.Add(totalMensajes)

	// 3. Lanzar todas las tareas concurrentemente
	for i := 0; i < totalMensajes; i++ {
		// La palabra `go` lanza la función en un hilo ligero (goroutine)
		// y el bucle for sigue inmediatamente a la siguiente iteración
		go simularPeticion(i, &wg)
	}

	// 4. Pausar la ejecución principal hasta que el contador del WaitGroup llegue a cero
	wg.Wait()

	// Calcular el tiempo transcurrido
	tiempoTotal := time.Since(inicio).Seconds()

	// En este ejemplo simple asumimos que todas fueron exitosas
	// Para contar éxitos de forma segura en concurrencia necesitarías canales (channels) o sync.Mutex
	exitosos := totalMensajes

	// 5. Mostrar estadísticas
	fmt.Printf("✅ Prueba finalizada en %.2f segundos.\n", tiempoTotal)
	fmt.Printf("📊 Mensajes exitosos: %d/%d\n", exitosos, totalMensajes)
	fmt.Printf("⚡ Rendimiento: %.2f peticiones/segundo\n", float64(totalMensajes)/tiempoTotal)
}
